### PR TITLE
Fix Booster read/write locale dependency

### PR DESCRIPTION
**Pull request by AlbertoEAF**
This request fixes the bug mentioned on the issue https://github.com/microsoft/LightGBM/issues/2890.

In C++, the process is by default is initialized to use the minimal "C" locale, however, in Java, when using the SWIG wrappers, this default is broken and the current machine locale is used instead.

Since during model read/write we rely on many C/C++ functions, some of which are 100% dependant on the locale, such as `std::stod()`, overriding the locale to a sane default is required.
Instead of setting the locale globally across the full process, as we might want to still read/write datasets according to the locale, a class to manage the locale context `LocaleContext` was created.
We **only** activate/deactivate that context (to force the "C" locale) **during model reads/writes**, to ensure our model format is independent of any locale changes in the machine or calling processes.

Motivation for this:

- In Python, the locale is always "C" by default, still as seen in the comment https://github.com/ninia/jep/issues/205#issuecomment-544078377, it is possible to break the model by changing locales.
- In Java, by using the SWIG interface, there isn't even a safety net like in Python, of using the "C" locale by default.

To avoid such issues, the fix should be done on the core library, so no bindings can be affected by locale issues, whatever happens, and this way the model format is always consistent.

### CODE REVIEW

**Comment by AlbertoEAF** [comment]
Hello @guolinke and @chivee, 

can you tell me if this looks good to you and if so, to approve/revise?

**Comment by guolinke** [comment]
@StrikerRUS any style issues?

File: include/LightGBM/utils/LocaleContext.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#ifndef __LOCALE_CONTEXT_H__
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]: Please add copyright message.
```
/*!
 * Copyright (c) 2020 Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See LICENSE file in the project root for license information.
 */
```

File: include/LightGBM/utils/LocaleContext.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#include <locale>
+
+/**
+ * Class to override the program global locale during this object lifetime.
+ * After the object is destroyed, the locale is returned to its original state.
+ *
+ * @warn This is not thread-safe.
+*/
+class LocaleContext
+{
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
{ should almost always be at the end of the previous line  [whitespace/braces] [4]
```

File: include/LightGBM/utils/LocaleContext.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+public:
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
public: should be indented +1 space inside class LocaleContext  [whitespace/indent] [3]
```

File: include/LightGBM/utils/LocaleContext.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
```

File: include/LightGBM/utils/LocaleContext.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  /**
+   * Override the current program global locale during this object lifetime.
+   *
+   * @param target_locale override the locale to this locale setting.
+   * @warn This is not thread-safe.
+   * @note This doesn't override cout, cerr, etc.
+   */
+  LocaleContext(const char* target_locale = "C") {
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
Constructors callable with one argument should be marked explicit.  [runtime/explicit] [5]
```

File: include/LightGBM/utils/LocaleContext.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    std::locale::global(std::locale(target_locale));
+  }
+
+  /**
+   * Restores the old global locale.
+   */
+  ~LocaleContext() {
+    std::locale::global(_saved_global_locale);
+  }
+private:
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
private: should be indented +1 space inside class LocaleContext  [whitespace/indent] [3]
```

File: include/LightGBM/utils/LocaleContext.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  std::locale _saved_global_locale; //!< Stores global locale at initialization.
+};
+
+#endif // __LOCALE_CONTEXT_H__
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
At least two spaces is best between code and comments  [whitespace/comments] [2]
```

File: include/LightGBM/utils/LocaleContext.h (hunk 1 above)
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]: rename file to `locale_context.h`

Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
#ifndef LIGHTGBM_LOCALE_CONTEXT_H_
```

File: include/LightGBM/utils/LocaleContext.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
+#ifndef __LOCALE_CONTEXT_H__
+#define __LOCALE_CONTEXT_H__
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
#define LIGHTGBM_LOCALE_CONTEXT_H_
```

File: include/LightGBM/utils/LocaleContext.h (hunk 7 above)
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
#endif  // LIGHTGBM_LOCALE_CONTEXT_H_
```

File: src/c_api.cpp [hunk 9]
```diff
@@ -17,6 +17,7 @@
 #include <LightGBM/utils/openmp_wrapper.h>
 #include <LightGBM/utils/random.h>
 #include <LightGBM/utils/threading.h>
+#include <LightGBM/utils/LocaleContext.h>
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Follow alphabetical order.

File: include/LightGBM/utils/LocaleContext.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <clocale>
+#include <locale>
+
+/**
+ * Class to override the program global locale during this object lifetime.
+ * After the object is destroyed, the locale is returned to its original state.
+ *
+ * @warn This is not thread-safe.
+*/
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
 */
```

File: include/LightGBM/utils/LocaleContext.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
+#ifndef __LOCALE_CONTEXT_H__
+#define __LOCALE_CONTEXT_H__
+
+#include <clocale>
+#include <locale>
+
+/**
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
/*!
```

File: include/LightGBM/utils/LocaleContext.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ * Class to override the program global locale during this object lifetime.
+ * After the object is destroyed, the locale is returned to its original state.
+ *
+ * @warn This is not thread-safe.
+*/
+class LocaleContext
+{
+public:
+
+  /**
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
  /*!
```

File: include/LightGBM/utils/LocaleContext.h [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+   *
+   * @param target_locale override the locale to this locale setting.
+   * @warn This is not thread-safe.
+   * @note This doesn't override cout, cerr, etc.
+   */
+  LocaleContext(const char* target_locale = "C") {
+    std::locale::global(std::locale(target_locale));
+  }
+
+  /**
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```suggestion
  /*!
```

File: include/LightGBM/utils/LocaleContext.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+   * Restores the old global locale.
+   */
+  ~LocaleContext() {
+    std::locale::global(_saved_global_locale);
+  }
+private:
+  std::locale _saved_global_locale; //!< Stores global locale at initialization.
```
Comment (StrikerRUS) [inline include/LightGBM/utils/LocaleContext.h]:
```
At least two spaces is best between code and comments  [whitespace/comments] [2]
```
```suggestion
  std::locale _saved_global_locale;  // Stores global locale at initialization.
```

Comment (AlbertoEAF) [reply include/LightGBM/utils/LocaleContext.h]: This was to enrich the doxygen-style comments for the member data.

I'm applying the suggestion anyway.

**Comment by AlbertoEAF** [comment]
Changed everything you requested @StrikerRUS :)

Anything else?

**Comment by AlbertoEAF** [comment]
Thanks for merging @StrikerRUS :)

**Pull request merged**